staging/src/k8s.io/apiserver/pkg/util/flag/globalflags.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flag
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	// libs that provide registration functions
+	"k8s.io/apiserver/pkg/util/logs"
+
+)
+
+// AddGlobalFlags explicitly registers flags that libraries (glog, verflag, etc.) register
+// against the global flagsets from "flag" and "github.com/spf13/pflag".
+// We do this in order to prevent unwanted flags from leaking into the kubernetes component's flagset.
+func AddGlobalFlags(fs *pflag.FlagSet) {
+	addGlogFlags(fs)
+	logs.AddFlags(fs)
+}
+
+// normalize replaces underscores with hyphens
+// we should always use hyphens instead of underscores when registering kubernetes component's flags
+func normalize(s string) string {
+	return strings.Replace(s, "_", "-", -1)
+}
+
+// register adds a flag to local that targets the Value associated with the Flag named globalName in global
+func register(global *flag.FlagSet, local *pflag.FlagSet, globalName string) {
```

Comment:
Since you're putting it in `apiserver/pkg/util/flag`, do you think we can expose this `register` function? (i.e. `register` -> `Register`.) This would be really useful especially when specific packages want to register their own flags. See https://github.com/kubernetes/kubernetes/pull/70204/commits/83c25d99810e96ea82e23effd3c6641e1be2ead1#diff-cd5ce47c38304a5e11f51dd5f9aab40aR90.

Reply:
Why own flag belong to the `global`?

Reply:
I'm not quite sure if I understand what you mean. Certain flags are registered through `flag` and they end up in the **global flag set** (`flag.CommandLine`): 
https://github.com/kubernetes/kubernetes/blob/521028ed52526abd832f8a6c9c5ec62340fccbb2/plugin/pkg/admission/defaulttolerationseconds/admission.go#L34-L40

The register function above is just a helper function that looks for flags in the first argument (global) based on the name, and put it into the second argument (local).

For example, there are additional custom generic flags for the `kube-apiserver` command such as the ones in `k8s.io/cloud-provider` and `k8s.io/apiserver/pkg/admission`. I'm thinking after merging this PR, we can create a file (called `customflags.go`? perhaps in `cmd/kube-apiserver/app/options/`) that manually registers the custom generic flags. With the register function exposed, we can just call `apiserver.Register` to extract those custom flags out from the global flag set. (We can do the same for the remaining *-controller-manager commands too.)

Reply:
BTW, can you help find out why `pull-kubernetes-integration ` CI failed when i move `globalflags.go` into `k8s.io/apiserver/pkg/util/flag` ?